### ISSUE TITLE
Batching transactions for P2P transmission

### ISSUE THREAD

**Issue by roman-khimov** [neo-project/neo#2365]
**Summary or problem description**
Transactions are distributed thoughout our P2P network via simple `Inv`-`GetData`-`Transaction` protocol where `Inv` announces transaction presence on the node to neighbors, `GetData` allows to request transactions from neighbor and `Transaction` is a transaction itself packed into P2P message. So `Inv` payload can carry up to 500 hashes, then `GetData` request can also contain up to 500 hashes but `Transaction` is always a transaction, one and only per message.

This makes our protocol somewhat asymmetric and inefficient, if I'm requesting 32 transactions from neighbor I'll get 32 small messages instead of one big message which adds some per-message networking overhead and doesn't allow for efficient compression.

There was an attempt to fix this in #1088, but it wasn't really effective because `Inv` messages there usually contained just one hash. This can be fixed if not already, but the most important thing here is whether the protocol allows for batching or not.

**Do you have any solution you want to propose?**
We can change `Transaction` message to `Transactions` and allow to send an array of transactions. It adds one byte overhead if we're to use this extended payload type for one transaction, but I think it's not a problem. It will still allow to send one hash per `Inv`/`GetData` and receive one transaction in return, but at the same time it will make it possible to respond with multiple transactions in one message for `GetData` requests containing multiple hashes. Nodes can then choose optimal behavior depending on implementation/configuration/other conditions (neo-go can send multiple hashes in `Inv` and request multiple transactions in `GetData`).

nspcc-dev/neo-go#1770 contains experimental implementation of this idea using a separate message command and [neo-bench](https://github.com/nspcc-dev/neo-bench/) shows that it can improve four node benchmark results up to 15% which is quite noticeable for such a simple change.

The same thing can be done to `Block` and `Extensible` payloads, though these are more rare and less critical.

**Neo Version**
- Neo 3

**Where in the software does this update applies to?**
- P2P (TCP)

**Comment by shargon** [comment]
I tried something like this in the past https://github.com/neo-project/neo/pull/1088

**Comment by roman-khimov** [comment]
Yep, it's mentioned above, but there was no visible improvement because the node still sent 1 hash per `Inv` (leading to one hash in `GetData` and one transaction sent in reply). My point is that it can be effective (15% is not nothing) if properly implemented (multiple hashes in `Inv`/`GetData` and subsequently multiple transactions packed into one mesage). And that's what can be measures with `neo-bench` running four nodes on one machine, on a real network optimized communication can bring even more improvement.

Either way, it's quite simple protocol change that doesn't cost a lot to have.

**Comment by shargon** [comment]
If you have some benchmarks it could help, my pr was closed because was not possible to prove this improvement.
